### PR TITLE
add support for model parallelism

### PR DESCRIPTION
**Pull request by karots123**

### CODE REVIEW

**Comment by karots123** [comment]
deepspeed+tp文档使用指令:
                TP_TencentPretrain/scripts/cconvert_llama_from_pytorch_checkpoint_to_megatron_checkpoint.py:
                将llama模型拆分成tp的格式,需要注意参数对齐：模型层数(layers_num),隐藏层维度(hidden_size),前馈层
                维度(feedforward_size),tp大小(tp_size)

                TP_TencentPretrain/scripts/convert_llama_from_megatron_checkpoint_to_pytorch_checkpoint.py:
                将deepspeed格式的tp模型组合成整模型:需要输入layers_num和tp_size,以及输入和输出文件
                
当前版本的项目文件兼容之前的TencentPretrain指令,运行tp需要加入--use_mp参数,以及指定--tp_size
例如此处指定了模型并行组size为4，gpu数量为8
运行指令参考:deepspeed pretrain.py --deepspeed --deepspeed_config models/deepspeed_config_bf16.json \
                     --dataset_path dataset_path.pt  --use_mp \
                     --spm_model_path sentencepiece.model \
                     --config_path models/llama/7b_config.json \
                     --output_model_path models/llama --scheduler cosine  --warmup 0.03 \
                     --world_size 8 --learning_rate 5e-5  --deepspeed_checkpoint_activations   \
                     --data_processor lm --total_steps 100000  \
                     --save_checkpoint_steps 10000 --batch_size 4 \
                     --report_steps 50 --tensor_model_parallel_size 4  \
                     --pretrained_model_path models/input_model 

2机每机8GPU预训练示例： 使用多机训练时需要先配置hostfile.txt文件，文件格式为：ip slots=gpu数 例如：
1.1.1.1 slots=8
2.2.2.2 slots=8
指定了模型并行组size为4，gpu数量为16
运行指令参考:deepspeed --hostfile=hostfile.txt pretrain.py --deepspeed --deepspeed_config models/deepspeed_config_bf16.json \
                     --dataset_path dataset_path.pt  --use_mp \
                     --spm_model_path sentencepiece.model \
                     --config_path models/llama/7b_config.json \
                     --output_model_path models/llama --scheduler cosine  --warmup 0.03 \
                     --world_size 16 --learning_rate 5e-5  --deepspeed_checkpoint_activations   \
                     --data_processor lm --total_steps 100000  \
                     --save_checkpoint_steps 10000 --batch_size 4 \
                     --report_steps 50 --tensor_model_parallel_size 4  \
                     --pretrained_model_path models/input_model

**Comment by karots123** [comment]
deepspeed+tp使用指令:

TP_TencentPretrain/scripts/cconvert_llama_from_pytorch_checkpoint_to_megatron_checkpoint.py:
将llama模型拆分成tp的格式,需要注意参数对齐：模型层数(layers_num),隐藏层维度(hidden_size),前馈层维度(feedforward_size),tp大小(tp_size)

TP_TencentPretrain/scripts/convert_llama_from_megatron_checkpoint_to_pytorch_checkpoint.py:
将deepspeed格式的tp模型组合成整模型:需要输入layers_num和tp_size,以及输入和输出文件
                
当前版本的项目文件兼容之前的TencentPretrain指令,运行tp需要加入--use_mp参数,以及指定--tp_size
例如此处指定了模型并行组size为4，gpu数量为8
运行指令参考:deepspeed pretrain.py --deepspeed --deepspeed_config models/deepspeed_config_bf16.json \
                     --dataset_path dataset_path.pt  --use_mp \
                     --spm_model_path sentencepiece.model \
                     --config_path models/llama/7b_config.json \
                     --output_model_path models/llama --scheduler cosine  --warmup 0.03 \
                     --world_size 8 --learning_rate 5e-5  --deepspeed_checkpoint_activations   \
                     --data_processor lm --total_steps 100000  \
                     --save_checkpoint_steps 10000 --batch_size 4 \
                     --report_steps 50 --tensor_model_parallel_size 4  \
                     --pretrained_model_path models/input_model 

2机每机8GPU预训练示例： 使用多机训练时需要先配置hostfile.txt文件，文件格式为：ip slots=gpu数 例如：
1.1.1.1 slots=8
2.2.2.2 slots=8
指定了模型并行组size为4，gpu数量为16
运行指令参考:deepspeed --hostfile=hostfile.txt pretrain.py --deepspeed --deepspeed_config models/deepspeed_config_bf16.json \
                     --dataset_path dataset_path.pt  --use_mp \
                     --spm_model_path sentencepiece.model \
                     --config_path models/llama/7b_config.json \
                     --output_model_path models/llama --scheduler cosine  --warmup 0.03 \
                     --world_size 16 --learning_rate 5e-5  --deepspeed_checkpoint_activations   \
                     --data_processor lm --total_steps 100000  \
                     --save_checkpoint_steps 10000 --batch_size 4 \
                     --report_steps 50 --tensor_model_parallel_size 4  \
                     --pretrained_model_path models/input_model

**Comment by wmpscc** [comment]
该版本针对 LLM 模型训练进行了修改：
### data_processor 修改
- `--data_processor` 参数增加 `llm_pretrain` 和 `llm_sft` 两种实现。`alpaca` 参数已删除，原功能由 `llm_sft` 支持。
    - `llm_pretrain` 对应LLM预训练阶段，训练数据采用jsonl格式（eg. `{"text": "大海的颜色上什么"}`），一行对应一条样本。
    - `llm_sft` 对应LLM指令微调阶段，训练数据采用jsonl格式（eg. `{"instruction": "大海",  "input": "大海的颜色上什么", "output": "蓝色"}`），一行对应一条样本。

#### 使用说明
`corpora/book_review.jsonl`文件内容
``` shell
> head corpora/book_review.jsonl
{"text": "我的入门扫盲教材，图文并茂，最初正是被本书的编写体例所吸引"}
{"text": "怎么会有这样的书。"}
...
```

数据预处理
``` shell
python3 preprocess.py --corpus_path corpora/book_review.jsonl \
--spm_model_path $LLaMA_7B_FOLDER/tokenizer.model --dataset_path dataset.pt \
--seq_length 256 --processes_num 32 --data_processor llm_pretrain
```

### 单机数据并行+张量并行
开启张量并行需指定 `--use_mp` 和 `--tensor_model_parallel_size` 参数。
#### zero-2
``` shell
deepspeed pretrain.py --deepspeed --deepspeed_config models/deepspeed_config.json \
--dataset_path dataset.pt --data_processor llm_pretrain \
--spm_model_path $LLaMA_7B_FOLDER/tokenizer.model \
--config_path models/llama/7b_config.json \
--output_model_path models_output/output_model.bin \
--learning_rate 5e-5 --scheduler cosine --batch_size 4 \
--total_steps 100000 --warmup 0.03 --report_steps 50 \
--save_checkpoint_steps 10000 \
--world_size 8 --deepspeed_checkpoint_activations
```

#### zero-2 + TP
``` shell
deepspeed pretrain.py --deepspeed --deepspeed_config models/deepspeed_config.json \
--dataset_path dataset.pt --data_processor llm_pretrain \
--spm_model_path $LLaMA_7B_FOLDER/tokenizer.model \
--config_path models/llama/7b_config.json \
--output_model_path models_output/output_model.bin \
--learning_rate 5e-5 --scheduler cosine --batch_size 4 \
--total_steps 100000 --warmup 0.03 --report_steps 50 \
--save_checkpoint_steps 10000 \
--world_size 8 --deepspeed_checkpoint_activations \
--use_mp --tensor_model_parallel_size 2
```

### 多机数据并行+张量并行
首先配置服务器信息，提供 `hostfile.txt` 文件，以2机16卡为例：
```
172.31.0.1 slots=8
172.31.0.2 slots=8
```

#### zero-2 + TP
``` shell
deepspeed --hostfile=hostfile.txt --master_addr $MASTER_ADDR pretrain.py --deepspeed --deepspeed_config models/deepspeed_config.json \
--dataset_path dataset.pt --data_processor llm_pretrain \
--spm_model_path $LLaMA_7B_FOLDER/tokenizer.model \
--config_path models/llama/7b_config.json \
--output_model_path models_output/output_model.bin \
--learning_rate 5e-5 --scheduler cosine --batch_size 4 \
--total_steps 100000 --warmup 0.03 --report_steps 50 \
--save_checkpoint_steps 10000 \
--world_size 8 --deepspeed_checkpoint_activations \
--use_mp --tensor_model_parallel_size 2
```

### 预训练权重转换
将llama模型拆分成megatron张量并行的格式。
- 注意对齐参数：模型层数(layers_num)，隐藏层维度(hidden_size)，前馈层维度(feedforward_size)，张量并行大小(tensor_model_parallel_size)。
- 模型参数信息可通过config文件获取，例如llama 7b的参数文件可查看`models/llama/7b_config.json`
```shell
python3 convert_llama_from_pytorch_checkpoint_to_megatron_checkpoint.py --input_model_path $LLaMA_7B_FOLDER/llama.bin \
--output_model_path $LLaMA_7B_FOLDER/llama_tp4 \
--layers_num 32 \
--hidden_size 4096 \
--feedforward_size 11008 \
--tensor_model_parallel_size 4
```

将合并张量并行权重
```shell
python3 scripts/convert_llama_from_megatron_checkpoint_to_pytorch_checkpoint.py --input_model_path $LLaMA_7B_FOLDER/llama_tp4 \
--output_model_path models/output_model \
--layers_num 32 \
--tensor_model_parallel_size 4
```

训练时增加 `--pretrained_model_path` 参数导入预训练权重
```shell
deepspeed --hostfile=hostfile.txt --master_addr $MASTER_ADDR pretrain.py --deepspeed --deepspeed_config models/deepspeed_config.json \
--dataset_path dataset.pt --data_processor llm_pretrain \
--spm_model_path $LLaMA_7B_FOLDER/tokenizer.model \
--config_path models/llama/7b_config.json \
--output_model_path models_output/output_model.bin \
--learning_rate 5e-5 --scheduler cosine --batch_size 4 \
--total_steps 100000 --warmup 0.03 --report_steps 50 \
--save_checkpoint_steps 10000 \
--world_size 8 --deepspeed_checkpoint_activations \
--use_mp --tensor_model_parallel_size 4 \
--pretrained_model_path $LLaMA_7B_FOLDER/llama_tp4
```

### BF16训练
修改`models/deepspeed_config.json`文件，将
``` json
  "fp16": {
    "enabled": true,
    "loss_scale": 0,
    "loss_scale_window": 1000,
    "hysteresis": 2,
    "min_loss_scale": 1
  }
```
替换为
```json
"bf16": {
   "enabled": true
 }
```
